nixos/modules/services/torrent/opentracker.nix

```diff
@@ -3,42 +3,237 @@
-        opentracker package to use
+    listenPort = mkOption {
+      type = types.int;
+      description = mdDoc ''
+        Port number to bind to.
+      '';
+      default = 6969;
+    };
+
+    listenAddress.ipv4 = mkOption {
+      type = types.nullOr types.string;
+      description = mdDoc ''
+        IPv4 address to bind to. Set to null to disable IPv4 server.
+      '';
+      default = "0.0.0.0";
+    };
+
+    listenAddress.ipv6 = mkOption {
+      type = types.nullOr types.string;
+      description = mdDoc ''
+        IPv6 address to bind to. Set to null to disable IPv6 server.
+      '';
+      default = "::";
+      example = "::1";
+    };
+
+    restrictStats = {
+      enable = mkOption {
+        type = types.bool;
+        description = mdDoc ''
+          Restrict access to tracker statistics.
+        '';
+        default = false;
+      };
+
+      allow.ipv4 = mkOption {
+        type = types.listOf types.string;
+        description = mdDoc ''
+          Allow these IPv4 addresses to access stats.
+        '';
+        default = [];
+      };
+
+      allow.ipv6 = mkOption {
+        type = types.listOf types.string;
+        description = mdDoc ''
+          Allow these IPv6 addresses to access stats.
+        '';
+        default = [];
+        example = [ "::1" ];
+      };
+    };
+
+    blacklist = mkOption {
+      type = types.nullOr (types.listOf types.string);
+      description = mdDoc ''
+        Do not track torrents with these hashes.
       '';
-      default = pkgs.opentracker;
-      defaultText = literalExpression "pkgs.opentracker";
+      default = null;
     };
 
-    extraOptions = mkOption {
-      type = types.separatedString " ";
-      description = lib.mdDoc ''
-        Configuration Arguments for opentracker
-        See https://erdgeist.org/arts/software/opentracker/ for all params
+    blacklistFile = mkOption {
+      type = types.nullOr types.path;
+      description = mdDoc ''
+        Do not track torrents with the hashes from this file.
+      '';
+      default = null;
+    };
+
+    whitelist = mkOption {
+      type = types.nullOr (types.listOf types.string);
+      description = mdDoc ''
+        Only track torrents with these hashes.
+      '';
+      default = null;
+    };
+
+    whitelistFile = mkOption {
+      type = types.nullOr types.path;
+      description = mdDoc ''
+        Only track torrents with the hashes from this file.
+      '';
+      default = null;
+    };
+
+    package = mkPackageOption pkgs "opentracker" {};
+
+    extraConfig = mkOption {
+      type = with types; attrsOf (oneOf [ string (listOf string) ]);
+      description = mdDoc ''
+        Configuration options for opentracker
+        See <https://erdgeist.org/gitweb/opentracker/tree/opentracker.conf.sample> for all params
+      '';
+      default = {};
+      example = literalExpression ''
+        {
+          "tracker.redirect_url" = "https://nixos.org/";
+        }
```

Comment:
This looks like a json style setting, why not use a json type for everything?

Reply:
Not really, it just key value pairs: <https://erdgeist.org/gitweb/opentracker/tree/opentracker.conf.sample>